src/main/scala/is/hail/utils/BinaryHeap.scala

```diff
@@ -0,0 +1,202 @@
+package is.hail.utils
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+class BinaryHeap[@specialized T: ClassTag](minimumCapacity: Int = 32) {
+  private var ts: Array[T] = new Array[T](minimumCapacity)
+  private var ranks: Array[Long] = new Array[Long](minimumCapacity)
+  private val m: mutable.Map[T, Int] = new mutable.HashMap()
+  private var next: Int = 0
+
+  def size: Int = next
+
+  def isEmpty: Boolean = next == 0
+  def nonEmpty: Boolean = next != 0
+
+  override def toString(): String =
+    s"values: ${ ts.slice(0, next): IndexedSeq[T] }; ranks: ${ ranks.slice(0, next): IndexedSeq[Long] }"
+
+  def insert(t: T, r: Long) {
+    if (m.contains(t))
+      throw new RuntimeException(s"key $t already exists with priority ${ ranks(m(t)) }, cannot add it again with priority $r")
+    maybeGrow()
+    put(next, t, r)
+    bubbleUp(next)
+    next += 1
+  }
+
+  private def emptyHeap() = new RuntimeException("heap is empty")
+
+  def max(): T =
+    if (next > 0)
+      ts(0)
+    else
+      throw emptyHeap()
+
+  def maxPriority(): Long =
+    if (next > 0)
+      ranks(0)
+    else
+      throw emptyHeap()
+
+  def extractMax(): T = {
+    val max = if (next > 0)
+      ts(0)
+    else
+      throw emptyHeap()
+
+    next -= 1
+    m -= max
+    if (next > 0) {
+      put(0, ts(next), ranks(next))
+      bubbleDown(0)
+      maybeShrink()
+    }
+    max
+  }
+
+  def getPriority(t: T): Long =
+    ranks(m(t))
+
+  def decreasePriorityTo(t: T, r: Long) {
+    val i = m(t)
+    assert(ranks(i) > r)
+    ranks(i) = r
+    bubbleDown(i)
+  }
+
+  def decreasePriority(t: T, f: (Long) => Long) {
+    val i = m(t)
+    val r = f(ranks(i))
+    assert(ranks(i) > r)
+    ranks(i) = r
+    bubbleDown(i)
+  }
+
+  def increasePriorityTo(t: T, r: Long) {
+    val i = m(t)
+    assert(ranks(i) < r)
+    ranks(i) = r
+    bubbleUp(i)
+  }
+
+  def increasePriority(t: T, f: (Long) => Long) {
+    val i = m(t)
+    val r = f(ranks(i))
+    assert(ranks(i) < r)
+    ranks(i) = r
+    bubbleDown(i)
```

Comment:
Bug: should be bubbleUp

Add regression test and fix in master (perhaps incorporating other tweaks herein too).